src/components/coin.ts

```diff
@@ -72,3 +91,82 @@ export const adjustCoinBalanceByUserId = async (userId: string, amount: number):
     amount
   );
 };
+
+/*
+  If (user, bonusType) doesn't exist, create row with current time as this bonusType log.
+  Otherwise, update last_granted to CURRENT_TIMESTAMP.
+*/
+export const updateUserBonusTableByUserId = async (userId: string, bonusType: BonusType): Promise<void> => {
+  const db = await openDB();
+  const bonusTypeInTable = bonusType as number;
+  await db.run(
+    `
+    INSERT INTO user_coin_bonus (user_id, bonus_type, last_granted) VALUES (?, ?, CURRENT_TIMESTAMP)
+    ON CONFLICT(user_id, bonus_type)
+    DO UPDATE SET last_granted = CURRENT_TIMESTAMP`,
+    userId,
+    bonusTypeInTable
+  );
+};
+
+/*
+  Get the time of the latest bonus applied to a user based on type
+*/
+export const latestBonusByUserId = async (userId: string, type: BonusType): Promise<UserCoinBonus[] | undefined> => {
+  const db = await openDB();
+  const res: UserCoinBonus[] | undefined = await db.get(
```

Comment:
if you use `get` I think it just returns a single result, so it wouldn't have type `UserCoinBonus[]`